### DISCUSSION TITLE
Cannot find OpenTK and Bonsai.Shaders in C# transform scripts

### DISCUSSION THREAD

**Discussion by ThetaS** [bonsai-rx/bonsai#969] [Q&A] [answered]
Hi Bonsai Team,

I came across this problem when I was reviewing the scripting tutorial as in: https://bonsai-rx.org/docs-wip/worksheets/scripting.html . My Bonsai command line prompt reads that it cannot identify Bonsai.Shaders and OpenTK as ref for the script (fig 1) despite both packages has been installed. 
![image](https://user-images.githubusercontent.com/32575952/178363377-9ecc63f5-d4be-48dc-a063-545cc2699365.png)

Fig 2 is my Bonsai workflow. I suppose Bonsai.Shaders and OpenTK are installed because Bonvision is functioning normally. I found that Bonvision does not support the latest version of OpenTK (ver 4.74), so my Bonvision plug-in is current supported by OpenTK ver 3.10. The node "RandomDot" is the custom C# transform node that I wrote (marked red with gridlines).
![image](https://user-images.githubusercontent.com/32575952/178364426-435c32b7-7360-497d-9d72-0372ed0cb7af.png)

I have tried multiple methods in resolving this issue including:
1) reinstall OpenTK, Bonsai.Shaders, Bonvision, and OpenCV.Net.
2) Check whether VS Code prompts notification for adding dependencies.
3) Repair .Net Core and Framework packages.

Additionally, sometimes it prompts another error when I attempt to save my workflow (fig 3).
![image](https://user-images.githubusercontent.com/32575952/178365789-416bb20b-b513-42da-826a-2a8046a25bcf.png)

The code in the C Sharp Transform is identical to the implementation in the tutorial above except the class name is "RandomDot". May I know what are the other methods that I can try to fix these problems? Thank you!

DT

**Comment by bruno-f-cruz** [comment, marked as answer]
Can you include you project files? Specifically, I suspect you are missing packages/references in your `Extensions.cproj` file.

**Reply by ThetaS** [reply]
Thanks for the quick response! I manually added the two packages reference into the Extensions.cproj and it works. Do we need to do this every time if we want to use packages other than Bonsai.core?

**Comment by bruno-f-cruz** [comment]
In general they need to be there yes. However, there is a little trick that
can help you with it. If you add the csharptransform node in front of a
node of a type that requires a specific package, bonsai will automatically
add that package as a reference to the .proj file.
Hope it helps :)

**Reply by ThetaS** [reply]
Hello, sorry I have another question related to this scripting tutorial. I was trying to split the big chunk of code inside the Process function into some small functions. But it seems like the program as now can only remember respawned dot positions from previous frame. Specifically, only respawned dots are displayed and other dots that move in range are lost tracked.  Currently I am confused about:

1) how source.Select() and Observable.Defer() work. Do they get called only once or every time when there is a frame update? 
2) Variable life cycles in the Defer function, how does it remember things from previous frame update?
3) Is there a way to use the the parameter "value" without using lambda expression?

Thank you!
[RandomDot.zip](https://github.com/bonsai-rx/bonsai/files/9105657/RandomDot.zip)
